File: tests/Parity/id-generator/ops.py
#!/usr/bin/env python3
"""Sinh dãy thao tác ngẫu nhiên cho HydrateIdGenerator.

Mỗi dòng là một JSON `{"op": ..., "args": [...]}`. Cả oracle Python lẫn subject
PHP chạy đúng dãy này rồi in giá trị trả về của từng thao tác; diff phải rỗng.

Dãy ngẫu nhiên tốt hơn assert viết tay ở chỗ nó chạm tới các tổ hợp push/pop
lồng nhau mà người viết test khó nghĩ ra — mà đó lại chính là nơi bộ đếm theo
scope dễ lệch.
"""
import json
import random
import sys

TAGS = ['div', 'span', 'p', 'ul', 'li', 'a', 'h1', 'code', 'button', 'img']
REACTIVE = ['if', 'switch', 'foreach', 'for', 'while']
BLOCKS = ['content', 'main', 'side_bar', 'workspace']


def generate(count: int, seed: int) -> list[dict]:
    rng = random.Random(seed)
    ops: list[dict] = []
    depth = 1  # scope gốc

    for _ in range(count):
        roll = rng.random()

        if roll < 0.16:
            ops.append({'op': 'nextElement', 'args': [rng.choice(TAGS)]})
        elif roll < 0.30:
            ops.append({'op': 'pushElement', 'args': [rng.choice(TAGS)]})
            depth += 1
        elif roll < 0.40:
            ops.append({'op': 'pushReactive', 'args': [rng.choice(REACTIVE)]})
            depth += 1
        elif roll < 0.47:
            ops.append({'op': 'pushCase', 'args': [rng.randint(1, 12)]})
            depth += 1
        elif roll < 0.53:
            blade = rng.choice(['$loop->index', '$i', None])
            ops.append({'op': 'pushLoopIteration', 'args': ['__loopIndex', blade]})
            depth += 1
        elif roll < 0.60:
            ops.append({'op': 'nextOutput', 'args': []})
        elif roll < 0.65:
            ops.append({'op': 'nextComponent', 'args': []})
        elif roll < 0.69:
            ops.append({'op': 'pushComponent', 'args': []})
            depth += 1
        elif roll < 0.72:
            ops.append({'op': 'nextBlockOutlet', 'args': []})
        elif roll < 0.76:
            ops.append({'op': 'nextYield', 'args': []})
        elif roll < 0.80:
            ops.append({'op': 'pushBlock', 'args': [rng.choice(BLOCKS)]})
            depth += 1
        elif roll < 0.84:
            ops.append({'op': 'depth', 'args': []})
        elif roll < 0.88:
            ops.append({'op': 'formatBladeHydrate', 'args': ['div-1-output-1']})
        elif roll < 0.91:
            ops.append({'op': 'formatJsId', 'args': ['div-1-output-1']})
        elif roll < 0.94 and depth > 8:
            # Thỉnh thoảng đặt lại — kiểm tra reset() dọn sạch mọi bộ đếm
            ops.append({'op': 'reset', 'args': []})
            depth = 1
        else:
            # Cố tình pop cả khi đã ở gốc: bản Python coi đó là no-op trả None,
            # bản PHP phải giống hệt chứ không được ném lỗi.
            ops.append({'op': 'popScope', 'args': []})
            depth = max(1, depth - 1)

    return ops


def main() -> int:
    count = int(sys.argv[1]) if len(sys.argv) > 1 else 5000
    seed = int(sys.argv[2]) if len(sys.argv) > 2 else 20260830

    for op in generate(count, seed):
        print(json.dumps(op, ensure_ascii=False, sort_keys=True))

    return 0


if __name__ == '__main__':
    sys.exit(main())

File: tests/Parity/id-generator/ops.php
#!/usr/bin/env php
<?php
/**
 * Sinh dãy thao tác cho HydrateIdGenerator.
 *
 * Mỗi dòng là một JSON `{"op": ..., "args": [...]}`. Subject chạy đúng dãy này
 * rồi in giá trị trả về của từng thao tác; diff với ảnh chụp golden phải rỗng.
 *
 * Dãy sinh máy tốt hơn assert viết tay ở chỗ nó chạm tới các tổ hợp push/pop
 * lồng nhau mà người viết test khó nghĩ ra — mà đó lại chính là nơi bộ đếm theo
 * scope dễ lệch.
 *
 * TẤT ĐỊNH theo seed bằng một LCG viết tay, KHÔNG dùng mt_rand: thuật toán của
 * mt_rand đổi theo phiên bản PHP (MT_RAND_PHP đã deprecated ở 8.4), mà golden
 * thì phải tái lập được sau nhiều năm. 12 dòng số học rẻ hơn một ảnh chụp chết.
 */
declare(strict_types=1);

const TAGS = ['div', 'span', 'p', 'ul', 'li', 'a', 'h1', 'code', 'button', 'img'];
const REACTIVE = ['if', 'switch', 'foreach', 'for', 'while'];
const BLOCKS = ['content', 'main', 'side_bar', 'workspace'];

$count = (int) ($argv[1] ?? 5000);
$seed = (int) ($argv[2] ?? 20260830);

// LCG numerical-recipes: state = state * 1664525 + 1013904223 (mod 2^32)
$state = $seed & 0xFFFFFFFF;
$rand = static function () use (&$state): float {
    $state = ($state * 1664525 + 1013904223) & 0xFFFFFFFF;

    return $state / 4294967296.0;
};
$pick = static function (array $xs) use ($rand) { return $xs[(int) ($rand() * count($xs))]; };

$depth = 1;   // scope gốc
$out = [];

for ($i = 0; $i < $count; $i++) {
    $roll = $rand();

    if ($roll < 0.16) {
        $out[] = ['args' => [$pick(TAGS)], 'op' => 'nextElement'];
    } elseif ($roll < 0.30) {
        $out[] = ['args' => [$pick(TAGS)], 'op' => 'pushElement'];
        $depth++;
    } elseif ($roll < 0.40) {
        $out[] = ['args' => [$pick(REACTIVE)], 'op' => 'pushReactive'];
        $depth++;
    } elseif ($roll < 0.47) {
        $out[] = ['args' => [1 + (int) ($rand() * 12)], 'op' => 'pushCase'];
        $depth++;
    } elseif ($roll < 0.53) {
        $blade = $pick(['$loop->index', '$i', null]);
        $out[] = ['args' => ['__loopIndex', $blade], 'op' => 'pushLoopIteration'];
        $depth++;
    } elseif ($roll < 0.60) {
        $out[] = ['args' => [], 'op' => 'nextOutput'];
    } elseif ($roll < 0.65) {
        $out[] = ['args' => [], 'op' => 'nextComponent'];
    } elseif ($roll < 0.69) {
        $out[] = ['args' => [], 'op' => 'pushComponent'];
        $depth++;
    } elseif ($roll < 0.72) {
        $out[] = ['args' => [], 'op' => 'nextBlockOutlet'];
    } elseif ($roll < 0.76) {
        $out[] = ['args' => [], 'op' => 'nextYield'];
    } elseif ($roll < 0.80) {
        $out[] = ['args' => [$pick(BLOCKS)], 'op' => 'pushBlock'];
        $depth++;
    } elseif ($roll < 0.84) {
        $out[] = ['args' => [], 'op' => 'depth'];
    } elseif ($roll < 0.91) {
        $out[] = ['args' => ['div-1-output-1'], 'op' => 'formatJsId'];
    } elseif ($roll < 0.94 && $depth > 8) {
        // Thỉnh thoảng đặt lại — kiểm tra reset() dọn sạch mọi bộ đếm.
        $out[] = ['args' => [], 'op' => 'reset'];
        $depth = 1;
    } else {
        // Cố tình pop cả khi đã ở gốc: phải là no-op trả null, không được ném.
        $out[] = ['args' => [], 'op' => 'popScope'];
        $depth = max(1, $depth - 1);
    }
}

foreach ($out as $op) {
    echo json_encode($op, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES), "\n";
}
